src/ng/rootScope.js

```diff
@@ -855,7 +857,7 @@ function $RootScopeProvider() {
             // Insanity Warning: scope depth-first traversal
             // yes, this code is a bit crazy, but it works and we have tests to prove it!
             // this piece should be kept in sync with the traversal in $broadcast
-            if (!(next = ((current.$$watchersCount && current.$$childHead) ||
+            if (!(next = ((!current.$$suspended && current.$$watchersCount && current.$$childHead) ||
```

Comment:
The comment above this line says " this piece should be kept in sync with the traversal in $broadcast".  So should it?  Sorry I'm not familiar enough with the code to know, but I thought it was worth raising just in case.

Reply:
Good spot! I think it makes sense to deviate in this case (i.e. suspending the scope should not prevent event propagation), but it is a good idea to reflect that in the comment to avoid future confusion).

Reply:
So what does it mean "suspending the scope should not prevent event propagation"?  So if a parent scope (not suspended) does a broadcast, any suspended descendants will receive that event and run the associated event code? 

If that's what you mean, I think I disagree.  I think it's perhaps a more consistent DX if such event receivers are suspended because it's a more absolute and complete "turning off" of the suspended descendants' functionality.  Partial suspension is weirder and harder to reason about, document, explain to other developers, IMO.  

Code in descendants that react to changes will be arbitrarily contained in event receivers and $watches/$onChanges, and without complete suspension I have to review all of it to make sure they understand which of it will be executed despite suspension and which won't.

Reply:
I think I agree with @poshest here. We should not pass broadcasts to suspended scopes.

Reply:
I think I disagree with both of you then :grin:

In my view `$suspend`/`$resume` is only related to change-detection (which in AngularJS is performed via `$digest` runs). Event propagation is an orthogonal concern and should be able to work independently from CD.

Suspending event propagation sounds to me a little like saying we will not trigger `$timeout`s if they come from a suspended scope (assuming we could determine the scope). Or not evaluate `$eval()`/`$evalAsync()` expressions.

IMO it would be way more confusing to stop propagating events.

Secondary reasons:
- By running a `$digest` once the scope is resumed, you should be able to get to the same state you would have gotten if the scope was not suspended (most of the time). I.e. you are able to "replay" the changes, because they are recorded in the app state. This is especially true if (as recommended) you watch expressions and handlers are idempotent.
- Events could be a handy way to trigger `$suspend`/`$resume` :grin:

Reply:
I think it makes sense to handle digest and events separately here, as scope events are quite different from the digest cycle, and could be useful as @gkalpak said, to control the suspension. If we want to give devs the option to shut down broadcasting, then we need something like this PR: https://github.com/angular/angular.js/pull/15877

Reply:
Well, I still feel that it's less confusing to have the descendants suspended in every respect, so yes, it would be a clearer API to include suspension of `$timeout`s, and `$eval()`/`$evalAsync()`s.  But you make it sound like implementation of that would be difficult. :P

By "By running a `$digest` once the scope is resumed", are you saying that the missed events would have, if not for being `$suspend`ed, affected a component's state and therefore the component will show the wrong state when it's `$resume`d? I wouldn't code a component like that.  But I guess there may be use cases I haven't thought of.

"Events could be a handy way to trigger $suspend/$resume". True!  Certainly if the original intent to suspend the current scope (not only descendent ones) comes to fruition, this would be crucial. But even then, I still prefer suspending only descendants.

While it's not my preference, I'm open to excluding $broadcasts from suspending scopes, and in that case I'd argue for

- changing the names to `$suspendDigests()` and `$resumeDigests()` and `$digestsAreSuspended()` to be clearer
- adding my proposed `$ancestorIsSuspended()` (`$ancestorDigestsAreSuspended()`!) so that developers could easily avoid reacting to `$broadcast`ed events on suspended scopes on a case-by-case basis.

Reply:
> Well, I still feel that it's less confusing to have the descendants suspended in every respect, so yes, it would be a clearer API to include suspension of $timeouts, and $eval()/$evalAsync()s. But you make it sound like implementation of that would be difficult. 

It's difficult / impossible, because ...

For all of the following, the scope and its descedants are excluded from any digests: $apply, $evalAsync, $timeout, $interval, $q, $http. This is because all of these trigger a digest on the $rootScope, not the current scope, so they will digest until they reach the suspended scope. 

So these functions are not bound to a specific scope, but to the change detection (digest cycle) in general.

timeout and interval have an option to skip the apply for that reason. Http could get one, but it's trickier: https://github.com/angular/angular.js/pull/12557. eval() doesn't trigger a digest.